// @kararty/dank-twitch-irc/dist/message/twitch-types/privmsg.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PrivmsgMessage = exports.parseActionAndMessage = void 0;
const channel_irc_message_1 = require("../irc/channel-irc-message");
const irc_message_1 = require("../irc/irc-message");
const tag_values_1 = require("../parser/tag-values");
// eslint-disable-next-line no-control-regex
const actionRegex = /^\u0001ACTION (.*)\u0001$/;
function parseActionAndMessage(trailingParameter) {
    const match = actionRegex.exec(trailingParameter);
    if (match == null) {
        return {
            isAction: false,
            message: trailingParameter,
        };
    }
    else {
        return {
            isAction: true,
            message: match[1],
        };
    }
}
exports.parseActionAndMessage = parseActionAndMessage;
class PrivmsgMessage extends channel_irc_message_1.ChannelIRCMessage {
    messageText;
    isAction;
    senderUsername;
    senderUserID;
    badgeInfo;
    badgeInfoRaw;
    badges;
    badgesRaw;
    bits;
    bitsRaw;
    color;
    colorRaw;
    displayName;
    emotes;
    emotesRaw;
    /**
     * Can be an array of Twitch AutoMod flagged words, for use in moderation and/or filtering purposes.
     *
     * If the `flags` tag is missing or of a unparseable format, this will be `undefined`. This is unlike most other
     * attributes which when missing or malformed will fail the message parsing. However since this attribute is
     * completely undocumented we cannot rely on the `flags` tag being stable, so this soft fallback is used instead.
     * While it will be a major version release if this attribute changes format in dank-twitch-irc, using this is still
     * at your own risk since it may suddenly contain unexpected data or turn `undefined` one day as
     * Twitch changes something. In short: **Use at your own risk** and make sure your
     * implementation can handle the case where this is `undefined`.
     */
    flags;
    /**
     * Twitch AutoMod raw flags string.
     *
     * If the `flags` tag is missing or of a unparseable format, this will be `undefined`. This is unlike most other
     * attributes which when missing or malformed will fail the message parsing. However since this attribute is
     * completely undocumented we cannot rely on the `flags` tag being stable, so this soft fallback is used instead.
     * In short, ensure your implementation can handle the case where this is `undefined` or is in
     * a format you don't expect.
     */
    flagsRaw;
    messageID;
    isMod;
    isModRaw;
    channelID;
    serverTimestamp;
    serverTimestampRaw;
    constructor(ircMessage) {
        super(ircMessage);
        const { isAction, message } = parseActionAndMessage((0, irc_message_1.requireParameter)(this, 1));
        this.messageText = message;
        this.isAction = isAction;
        this.senderUsername = (0, irc_message_1.requireNickname)(this);
        const tagParser = (0, tag_values_1.tagParserFor)(this.ircTags);
        this.channelID = tagParser.requireString("room-id");
        this.senderUserID = tagParser.requireString("user-id");
        this.badgeInfo = tagParser.requireBadges("badge-info");
        this.badgeInfoRaw = tagParser.requireString("badge-info");
        this.badges = tagParser.requireBadges("badges");
        this.badgesRaw = tagParser.requireString("badges");
        this.bits = tagParser.getInt("bits");
        this.bitsRaw = tagParser.getString("bits");
        this.color = tagParser.getColor("color");
        this.colorRaw = tagParser.requireString("color");
        // trim: Twitch workaround for unsanitized data, see https://github.com/robotty/dank-twitch-irc/issues/33
        this.displayName = tagParser.requireString("display-name").trim();
        this.emotes = tagParser.requireEmotes("emotes", this.messageText);
        this.emotesRaw = tagParser.requireString("emotes");
        this.flags = tagParser.getFlags("flags", this.messageText);
        this.flagsRaw = tagParser.getString("flags");
        this.messageID = tagParser.requireString("id");
        this.isMod = tagParser.requireBoolean("mod");
        this.isModRaw = tagParser.requireString("mod");
        this.serverTimestamp = tagParser.requireTimestamp("tmi-sent-ts");
        this.serverTimestampRaw = tagParser.requireString("tmi-sent-ts");
    }
    /**
     * Extracts a plain object only containing the fields defined by the
     * {@link PrivmsgUserState} interface.
     */
    extractUserState() {
        return {
            badgeInfo: this.badgeInfo,
            badgeInfoRaw: this.badgeInfoRaw,
            badges: this.badges,
            badgesRaw: this.badgesRaw,
            color: this.color,
            colorRaw: this.colorRaw,
            displayName: this.displayName,
            isMod: this.isMod,
            isModRaw: this.isModRaw,
        };
    }
    isCheer() {
        return this.bits != null;
    }
}
exports.PrivmsgMessage = PrivmsgMessage;
