// @kararty/dank-twitch-irc/dist/message/twitch-types/usernotice.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UsernoticeMessage = exports.extractEventParams = exports.getCamelCasedName = void 0;
const camelCase = require("lodash.camelcase");
const channel_irc_message_1 = require("../irc/channel-irc-message");
const irc_message_1 = require("../irc/irc-message");
const tag_values_1 = require("../parser/tag-values");
const convertersMap = {
    "msg-param-cumulative-months": tag_values_1.convertToInt,
    "msg-param-gift-months": tag_values_1.convertToInt,
    "msg-param-sender-count": tag_values_1.convertToInt,
    "msg-param-months": tag_values_1.convertToInt,
    "msg-param-promo-gift-total": tag_values_1.convertToInt,
    "msg-param-should-share-streak": tag_values_1.convertToBoolean,
    "msg-param-streak-months": tag_values_1.convertToInt,
    "msg-param-viewerCount": tag_values_1.convertToInt,
    "msg-param-threshold": tag_values_1.convertToInt,
    "msg-param-mass-gift-count": tag_values_1.convertToInt,
    "msg-param-origin-id": tag_values_1.convertToString,
    "msg-param-sub-plan": tag_values_1.convertToString,
};
function getCamelCasedName(tagKey) {
    let newKey = tagKey;
    // remove the leading msg-param-
    newKey = newKey.substring(10);
    // camel case
    newKey = camelCase(newKey);
    // convert somethingId to somethingID
    newKey = newKey.replace(/Id$/g, "ID");
    // To be consistent with the rest of the library,
    // don't camelcase username as userName
    newKey = newKey.replace(/([uU])serName/g, "$1sername");
    return newKey;
}
exports.getCamelCasedName = getCamelCasedName;
function extractEventParams(tags) {
    const params = {};
    // converts all msg-param-* tags into a new "params" object where keys are camelCased
    // and boolean/integer tags are parsed (including a identically named "Raw" property).
    // e.g. msg-param-should-share-streak would become
    // shouldShareStreak: true
    // shouldShareStreakRaw: '1'
    for (const tagKey of Object.keys(tags)) {
        if (!tagKey.startsWith("msg-param-")) {
            continue;
        }
        const newKey = getCamelCasedName(tagKey);
        const converter = convertersMap[tagKey];
        if (converter != null) {
            params[newKey] = (0, tag_values_1.requireData)(tags, tagKey, converter);
            params[newKey + "Raw"] = (0, tag_values_1.requireData)(tags, tagKey, tag_values_1.convertToString);
        }
        else {
            params[newKey] = (0, tag_values_1.requireData)(tags, tagKey, tag_values_1.convertToString);
        }
    }
    return params;
}
exports.extractEventParams = extractEventParams;
class UsernoticeMessage extends channel_irc_message_1.ChannelIRCMessage {
    channelID;
    messageText;
    systemMessage;
    /** sub, resub, subgift, etc... */
    messageTypeID;
    senderUsername;
    senderUserID;
    badgeInfo;
    badgeInfoRaw;
    badges;
    badgesRaw;
    bits;
    bitsRaw;
    color;
    colorRaw;
    displayName;
    emotes;
    emotesRaw;
    /**
     * Can be an array of Twitch AutoMod flagged words, for use in moderation and/or filtering purposes.
     *
     * If the `flags` tag is missing or of a unparseable format, this will be `undefined`. This is unlike most other
     * attributes which when missing or malformed will fail the message parsing. However since this attribute is
     * completely undocumented we cannot rely on the `flags` tag being stable, so this soft fallback is used instead.
     * While it will be a major version release if this attribute changes format in dank-twitch-irc, using this is still
     * at your own risk since it may suddenly contain unexpected data or turn `undefined` one day as
     * Twitch changes something. In short: **Use at your own risk** and make sure your
     * implementation can handle the case where this is `undefined`.
     */
    flags;
    /**
     * Twitch AutoMod raw flags string.
     *
     * If the `flags` tag is missing or of a unparseable format, this will be `undefined`. This is unlike most other
     * attributes which when missing or malformed will fail the message parsing. However since this attribute is
     * completely undocumented we cannot rely on the `flags` tag being stable, so this soft fallback is used instead.
     * In short, ensure your implementation can handle the case where this is `undefined` or is in
     * a format you don't expect.
     */
    flagsRaw;
    messageID;
    isMod;
    isModRaw;
    serverTimestamp;
    serverTimestampRaw;
    eventParams;
    constructor(message) {
        super(message);
        this.messageText = (0, irc_message_1.getParameter)(this, 1);
        const tagParser = (0, tag_values_1.tagParserFor)(this.ircTags);
        this.channelID = tagParser.requireString("room-id");
        this.systemMessage = tagParser.requireString("system-msg");
        this.messageTypeID = tagParser.requireString("msg-id");
        this.senderUsername = tagParser.requireString("login");
        this.senderUserID = tagParser.requireString("user-id");
        this.badgeInfo = tagParser.requireBadges("badge-info");
        this.badgeInfoRaw = tagParser.requireString("badge-info");
        this.badges = tagParser.requireBadges("badges");
        this.badgesRaw = tagParser.requireString("badges");
        this.bits = tagParser.getInt("bits");
        this.bitsRaw = tagParser.getString("bits");
        this.color = tagParser.getColor("color");
        this.colorRaw = tagParser.requireString("color");
        // trim: Twitch workaround for unsanitized data, see https://github.com/robotty/dank-twitch-irc/issues/33
        this.displayName = tagParser.requireString("display-name").trim();
        if (this.messageText != null) {
            this.emotes = tagParser.requireEmotes("emotes", this.messageText);
            this.flags = tagParser.getFlags("flags", this.messageText);
        }
        else {
            this.emotes = [];
        }
        this.emotesRaw = tagParser.requireString("emotes");
        this.flagsRaw = tagParser.getString("flags");
        this.messageID = tagParser.requireString("id");
        this.isMod = tagParser.requireBoolean("mod");
        this.isModRaw = tagParser.requireString("mod");
        this.serverTimestamp = tagParser.requireTimestamp("tmi-sent-ts");
        this.serverTimestampRaw = tagParser.requireString("tmi-sent-ts");
        this.eventParams = extractEventParams(this.ircTags);
    }
    isCheer() {
        return this.bits != null;
    }
    isSub() {
        return this.messageTypeID === "sub";
    }
    isResub() {
        return this.messageTypeID === "resub";
    }
    isRaid() {
        return this.messageTypeID === "raid";
    }
    isSubgift() {
        return this.messageTypeID === "subgift";
    }
    isMassSubgift() {
        return this.messageTypeID === "submysterygift";
    }
    isAnonSubgift() {
        return this.messageTypeID === "anonsubgift";
    }
    isAnonGiftPaidUpgrade() {
        return this.messageTypeID === "anongiftpaidupgrade";
    }
    isGiftPaidUpgrade() {
        return this.messageTypeID === "giftpaidupgrade";
    }
    isRitual() {
        return this.messageTypeID === "ritual";
    }
    isBitsBadgeTier() {
        return this.messageTypeID === "bitsbadgetier";
    }
}
exports.UsernoticeMessage = UsernoticeMessage;
